### PR TITLE
Ipc4: container size support in format conversion

### PR DESCRIPTION
**Pull request by RanderWang**
Ipc4 stream have container size and valid sample size.
    In windows system, container size is normally 32bit
    and valid sample size is 16bit or 24bit rely on system
    setting. Currently host and dai component are included
    by copier module to manage gateway. These two components
    doesn't support container size and all the code is designed
    for the case that container size is always equal to valid
    sample size.

With this patch HDA playback works with default windows audio player in format of 16bit/48000HZ/2CH.  24bit will be added later and it is a little tricky since it is really S24_3LE

### CODE REVIEW

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
Looks like some build failures
```
[ 92%] Building C object test/cmocka/src/math/arithmetic/CMakeFiles/base2_logarithm.dir/__/__/__/__/__/src/math/base2log.c.o
/home/runner/work/sof/sof/test/cmocka/src/audio/volume/volume_process.c: In function ‘test_audio_vol’:
/home/runner/work/sof/sof/test/cmocka/src/audio/volume/volume_process.c:377:2: error: enumeration value ‘SOF_IPC_FRAME_S24_3LE’ not handled in switch [-Werror=switch]
  377 |  switch (vol_state->sink->stream.frame_fmt) {
      |  ^~~~~~
[ 93%] Linking C executable base2_logarithm
```

File: src/ipc/ipc4/helper.c [hunk 1]
```diff
@@ -201,7 +201,7 @@ static struct comp_buffer *ipc4_create_buffer(struct comp_dev *src, struct comp_
 	src_cfg = (struct ipc4_base_module_cfg *)comp_get_drvdata(src);
 	sink_cfg = (struct ipc4_base_module_cfg *)comp_get_drvdata(sink);
 
-	buf_size = MAX(src_cfg->obs, sink_cfg->ibs);
+	buf_size = src_cfg->obs * 2;
```
Comment (lgirdwood) [inline src/ipc/ipc4/helper.c:204]: Can we also have a comment here explaining the logic.

Comment (RanderWang) [reply src/ipc/ipc4/helper.c:204]: sure

**Comment by lgirdwood** [comment]
Still build errors building the cmocka UTs. Can you locally reproduce using ```./scripts/run-mocks.sh``` ?
```
[ 97%] Building C object test/cmocka/src/audio/selector/CMakeFiles/selector_test.dir/selector_test.c.o
[ 97%] Built target demux_copy
[ 97%] Building C object test/cmocka/src/audio/eq_iir/CMakeFiles/eq_iir_process.dir/eq_iir_process.c.o
/home/runner/work/sof/sof/test/cmocka/src/audio/selector/selector_test.c: In function ‘test_audio_sel’:
/home/runner/work/sof/sof/test/cmocka/src/audio/selector/selector_test.c:270:2: error: enumeration value ‘SOF_IPC_FRAME_S24_3LE’ not handled in switch [-Werror=switch]
  270 |  switch (cd->source_format) {
      |  ^~~~~~
cc1: all warnings being treated as errors
make[2]: *** [test/cmocka/src/audio/selector/CMakeFiles/selector_test.dir/build.make:77: test/cmocka/src/audio/selector/CMakeFiles/selector_test.dir/selector_test.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1361: test/cmocka/src/audio/selector/CMakeFiles/selector_test.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by RanderWang** [comment]
updated, thanks!

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**